common/src/main/java/com/aliyun/emr/rss/common/network/server/FileManagedBuffers.java

```diff
@@ -66,23 +62,27 @@ public boolean hasAlreadyRead(int chunkIndex) {
     }
   }
 
-  @Override
-  public ManagedBuffer next() {
-    // This method is only used to clear the Managed Buffer when streamManager.connectionTerminated
-    // is called.
+  public ManagedBuffer chunk(int chunkIndex, int offset, int len) {
     synchronized (chunkTracker) {
-      index = chunkTracker.nextClearBit(index);
+      chunkTracker.set(chunkIndex, true);
     }
-    assert index < numChunks;
-    return chunk(index);
+    // offset of the beginning of the chunk in the file
+    final long chunkOffset = offsets[chunkIndex];
+    final long chunkLength = offsets[chunkIndex + 1] - chunkOffset;
+    assert offset < chunkLength;
+    long length = Math.min(chunkLength - offset, len);
+    if (len + offset >= chunkLength) {
+      synchronized (chunkTracker) {
+        chunkTracker.set(chunkIndex);
+      }
+      if (chunkIndex == numChunks - 1) {
+        fullyRead = true;
```

Comment:
In current logic seems we have not considered the case that some chunks read partial data?

Reply:
It's under the condition len + offset >= chunkLength, which means the current chunk is fully read.